\begin{titlepage}

%\thispagestyle{empty}

\newgeometry{left=6cm,bottom=0.5cm, top=0.25cm, right=1cm}

\tikz[remember picture,overlay] \node[opacity=1,inner sep=0pt] at (2.2mm,-165mm){\includegraphics{Fond1.png}}; % Fond changeable 

% fonte sans empattement pour la page de titre
\fontfamily{fvs}\fontseries{m}\selectfont

\color{white}

 
%**************************************************************
%********************  LOGO  DE  POLYTECH  ********************
%****** CHANGER L'IMAGE POUR UN AUTRE POLYTECH QUE SACLAY *****
%* VOIR LES LOGOS DISPONIBLES, ET REMPLACER LE NOM CI-DESSOUS *
%**************************************************************

\vspace{-5mm} % à ajuster en fonction de la hauteur du logo
               % décalera le texte en dessous
\flushright \includegraphics[width=70mm]{logo-ULB.jpg} % nom du logo du polytech

\vspace{-23mm}

\flushleft \includegraphics[height=19mm]{Images/VUB_RGB.png} %\caption{cghj}




%*****************************************************
%******************** TITRE **************************
%*****************************************************
\flushright
\vspace{15mm} % à régler éventuellement
\color{BleuFonce}
\fontfamily{cmss}\fontseries{m}\fontsize{22}{26}\selectfont
Review of `Comparing sequences without using alignments: application to
HIV/SIV subtyping'

\normalsize
\color{black}
%*****************************************************

%\fontfamily{fvs}\fontseries{m}\fontsize{8}{12}\selectfont

\vspace{1.5cm}
\normalsize

\textbf{Supervised by Prof. Mathieu Defrance (\textit{Machine Learning Group - ULB}) \\ \& Prof. Wim Vranken (\textit{Bioengineering Sciences Department - VUB})}

\vspace{15mm}

Master's in Computer Science\\  %% nom spécialité / enseignement

\vspace{15mm}

\Large 
{\color{BleuFonce} \textbf{Tom\'as GUIJA VALIENTE}}\\
{\color{BleuFonce} \textbf{Benjamin OBERTHÜR}}

\vspace{20mm}

\centering

\hrule
\begin{abstract}
    This article presents a review of an innovative method for alignment-free sequence comparison in the field of bioinformatics. The method, introduced in a relatively early published article, enables the identification of evolutionary relationships between sequences and highlights divisions into subtypes. Despite the absence of a description for the cluster merging method in the original article, the remaining parts of the algorithm were successfully re-implemented, demonstrating that the experiments were not environment-dependent and consistently produced reliable results. Furthermore, comparisons with more recent methods were discussed, acknowledging the possibility of newer approaches offering improved efficiency, runtime, and accuracy.\\

    The review emphasizes the significance of the alignment-free sequence comparison approach, particularly in addressing challenges posed by sequence alterations and the limitations of classic alignment methods. By leveraging the concept of $N$-words (or k-mers in today's scientific literature), dissimilarity matrices and clustering trees were constructed, providing insights into the evolutionary relationships among the sequences.\\

    The article concludes by highlighting the potential of the reviewed method as a valuable tool in phylogenetic analysis. It suggests avenues for further research, including exploring alternative clustering techniques and investigating the impact of various distance measures for enhanced accuracy. Overall, this review contributes to the understanding of alignment-free sequence comparison methods and their applications in studying evolutionary relationships among biological sequences.
\end{abstract}
\vspace*{3mm}
\hrule

\end{titlepage}